AllASPFuncs.py

```diff
@@ -70,107 +70,134 @@
-        # brightness.
-        imcopy = np.copy(image)
-        imcopy[imcopy <= 0] = 1e-10
-        bins = [-np.inf]
-        if len(percentiles) == 0:
-            percentiles = [45, 90]
-        bins.extend(np.nanpercentile(np.log(imcopy[np.where(np.log10(imcopy)>-10)]), percentiles))
-        bins.append(np.inf)
-
-        a = np.digitize(np.log(np.copy(imcopy)),bins)
-        xes = []
-        ys = []
-
-        xvals = x
-        yvals = y
-        yvals = np.rint(yvals).astype(int)
-        xvals = np.rint(xvals).astype(int)
-        for xindex in xvals:
-            x = xindex + 1
-            for yindex in yvals:
-                y = yindex + 1
-                num = int(a[x][y])
-                if num == 0:
-                    pass
-                elif num == 1:
-                    xes.append(y)
-                    ys.append(x)
-                else:
-                    xx = np.linspace(x - 0.6, x + 0.6, num+2)[1:-1]
-                    yy = np.linspace(y - 0.6, y + 0.6, num+2)[1:-1]
-                    X, Y = np.meshgrid(xx, yy)
-                    ys.extend(list(X.flatten()))
-                    xes.extend(list(Y.flatten()))
+    xx, yy = np.meshgrid(x+1, y+1)
+    xx = xx.flatten()
+    yy = yy.flatten()
+    Lager.debug(f'Built a grid with {np.size(xx)} points')
 
-        xx = np.array(xes)
-        yy = np.array(ys)
+    result = wcs.toWorld(xx, yy, units='deg')
+    ra_grid = result[0]
+    dec_grid = result[1]
+    return ra_grid, dec_grid
 
-    elif image is not None and makecontourGrid:
-        Lager.debug('USING CONTOUR GRID')
-        xx, yy = contourGrid(image)
-        xx = np.array(xx)
-        yy = np.array(yy)
 
-    else:
-        xx, yy = np.meshgrid(x+1, y+1)
+def make_adaptive_grid(ra_center, dec_center, wcs,
+                       image, percentiles=[45, 90], subsize=9):
     '''
-        subsize = 8 #Taking a smaller square inside the image to fit on
-        difference = int((size - subsize)/2)
+    Construct an "adaptive grid" which allocates model grid points to model
+    the background galaxy according to the brightness of the image.
 
-        spacing = 1.0
-        x = np.arange(difference, subsize+difference, spacing)
-        y = np.arange(difference, subsize+difference, spacing)
+    Inputs:
+    ra_center, dec_center: floats, coordinate center of the image
+    wcs: the WCS of the image, currently a galsim.fitswcs.AstropyWCS object
+    image: 2D numpy array of floats of shape (size x size), the image to build
+    the grid on.
+    percentiles: list of floats, the percentiles to use to bin the image. The
+                more bins, the more possible grid points could be placed in
+                that pixel. For instance, say if you had bins [45, 90],
+                as is default. A pixel that lies in the 30th percentile for
+                brightness would get 1 point. A pixel at the 50th percentiile
+                for brightness would get a 2x2 grid of points.
```

Comment:
Is the number points a pixel gets determined by the order of the percentiles?  

[45, 90] -> [1x1, 2x2]
[10, 25, 50, 75, 90] -> [1x1, 2x2, 3x3, 4x4, 5x5]

What happens if the list of floats is not in ascending order?
[45, 90, 10] -> ???

How many images do you need to provide constraints on a 5x5 subpixel grid?